src/lib.rs

```diff
@@ -1,20 +1,152 @@
+//! affect your API. This way your API can stabilize before all features in this crate are fully
+//! stable and you still can use all of them.
+//!
+//! ## Crate features
+//!
+//! * `std` - enables the standard library, on by default.
+//! * `alloc` - enables features that require allocation such as decoding into `Vec<u8>`, implied
+//! by `std`.
+//! * `newer-rust-version` - enables Rust version detection and thus newer features, may add
+//!                          dependency on a feature detection crate to reduce compile times. This
+//!                          feature is expected to do nothing once the native detection is in Rust
+//!                          and our MSRV is at least that version. We may also remove the feature
+//!                          gate in 2.0 with semver trick once that happens.
+//!
+//! ## MSRV policy
+//!
+//! The MSRV of the crate is currently 1.63.0 and we don't intend to bump it until the newer Rust
+//! version is at least two years old and also included in Debian stable (1.63 is in Debian 12 at
+//! the moment).
+//!
+//! Note though that the dependencies may have looser policy. This is not considered breaking/wrong
+//! - you would just need to pin them in `Cargo.lock` (not `.toml`).
 
-#![cfg_attr(all(not(test), not(feature = "std")), no_std)]
+#![no_std]
 // Experimental features we need.
 #![cfg_attr(docsrs, feature(doc_cfg))]
 #![cfg_attr(docsrs, feature(doc_auto_cfg))]
 // Coding conventions
 #![warn(missing_docs)]
 
+#[cfg(feature = "std")]
+extern crate std;
+
 #[cfg(feature = "alloc")]
 extern crate alloc;
+
+mod error;
+mod iter;
+
+#[cfg(feature = "alloc")]
+use alloc::vec::Vec;
+
+use crate::iter::HexToBytesIter;
+
+#[rustfmt::skip]                // Keep public re-exports separate.
+#[doc(inline)]
+pub use self::error::{DecodeFixedSizedBytesError, InvalidCharError, InvalidLengthError};
+#[cfg(feature = "alloc")]
+pub use self::error::{DecodeDynSizedBytesError, OddLengthStringError};
+
+/// Decodes a hex string with variable length.
+///
+/// The length of the returned `Vec` is determined by the length of the input, meaning all even
+/// lengths of the input string are allowed. If you know the required length at compile time using
+/// [`decode_fixed_sized`] is most likely a better choice.
+///
+/// # Errors
+///
+/// Errors if `hex` contains invalid characters or doesn't have even length.
+#[cfg(feature = "alloc")]
+pub fn decode_dyn_sized(hex: &str) -> Result<Vec<u8>, DecodeDynSizedBytesError> {
+    Ok(HexToBytesIter::new(hex)?.drain_to_vec()?)
+}
+
+/// Decodes a hex string with expected length kown at compile time.
+///
+/// If you don't know the required length at compile time you need to use [`decode_dyn_sized`]
+/// instead.
+///
+/// # Errors
+///
+/// Errors if `hex` contains invalid characters or has incorrect length. (Should be `N * 2`.)
+pub fn decode_fixed_sized<const N: usize>(hex: &str) -> Result<[u8; N], DecodeFixedSizedBytesError> {
```

Comment:
In 040a20e1787922dbe483800406ad3c81ef07de82:

`sized` means something in Rust that has nothing to do with this function. I guess you mean `fixed_size`, but then this is in contradiction with the commit message where you say that maybe we'll also want a method to decode fixed-size objects to `Vec<u8>`.

`decode_array` was fine. `decode_fixed_size` would also be ok, even though it's longer and less descriptive and might conflict with future functions.

Reply:
I definitely didn't mean `size` since that'd sound like we're decoding some kind of size.

Also it's not really that far off Rut terminology - arrays are `Sized` and whatever is backing a `Vec` is not.
There is no contradiction - if a function is needed we can add it later. If `Box::write` was stable I might've just ignored this and tell people to use it with this function since it should optimize well.

Reply:
If we use `fixed_size` it sounds like we're decoding a fixed-size object. It does not sound like we're decoding a `size`.

And `Vec`s are `Sized`. This terminology is confusing.